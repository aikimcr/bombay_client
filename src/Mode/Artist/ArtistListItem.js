import React, { useState } from 'react';

import './ArtistList.scss';

import Artist from './Artist';
import FormModal from '../../Modal/FormModal';

function ArtistListItem(props) {
    const [showEdit, setShowEdit] = useState(false);
    const [displayName, setDisplayName] = useState(props.artist.get('name'));

    async function updateArtist(modelDef) {
        await props.artist.set(modelDef).save();
        setDisplayName(props.artist.get('name'));
    }

    return (
        <React.Fragment>
            <li className="card" onClick={() => setShowEdit(true)}>
                <div className="header">Artist</div>
                <div className="name">{displayName}</div>
                <div className="details"></div>
            </li>
            <FormModal 
                title="Edit Artist"
                onClose={() => setShowEdit(false)}
                onSubmit={updateArtist}
                open={showEdit}
            >
                <Artist artist={props.artist} />
            </FormModal>
        </React.Fragment>
    )
}

export default ArtistListItem;