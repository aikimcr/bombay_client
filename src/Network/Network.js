// Manage XHR request calls.  This is the low-level stuff.
// import serverConfig from './serverConfig';
const serverConfig = {};

const serverConfigKeys = {
  serverProtocol: 'REACT_APP_SERVER_PROTOCOL',
  serverHost: 'REACT_APP_SERVER_HOST',
  serverBasePath: 'REACT_APP_SERVER_BASE_PATH',
  serverPort: 'REACT_APP_SERVER_PORT',
}

for (const key in serverConfigKeys) {
  if (process.env.hasOwnProperty(serverConfigKeys[key])) {
    const value = process.env[serverConfigKeys[key]];

    switch(value) {
      case 'null':
      case 'none':
      case 'empty':
        serverConfig[key] = null;
        break;

      default:
        serverConfig[key] = value;
    }
  }
}

function setConfigOption(key, defaultConfig) {
  return serverConfig.hasOwnProperty(key) ? serverConfig[key] : defaultConfig;
}

export const serverProtocol = setConfigOption('serverProtocol', 'http');
export const serverHost = setConfigOption('serverHost', 'localhost');
export const serverBasePath = setConfigOption('serverBasePath', '');
export const serverPort = setConfigOption('serverPort', 2001);

console.log(`Server Base URL: ${prepareURLFromArgs('')}`);

function getStandardHeaders(includeContentType = true) {
  const result = {};
  if (includeContentType) result['content-type'] = 'application/json';

  const token = localStorage.getItem('jwttoken')

  if (token) {
    result.Authorization = `Bearer ${token}`;
  }

  return result;
}

export function normalizeAndJoinPath(...pathParts) {
  const newPath = pathParts.reduce((memo, part) => {
    if (part === undefined) {
      throw new Error(`Unable to normalize path with parts: "${pathParts.join(', ')}"`);
    }

    let newPart = memo + '/' + part;
    newPart = newPart.replace(/\/\/+/g, '/');

    if (part !== '/') {
      if (newPart.length > 1) {
        newPart = newPart.replace(/\/+$/, '');
      }
    }

    return newPart;
  }, '/');

  return newPath;
}

export function buildURL(args = {}) {
  let { hostname, path, protocol, port } = {
    hostname: serverHost,
    path: '/',
    protocol: serverProtocol,
    port: serverPort,
    ...args,
  }

  hostname = hostname.replace(/^\/+/, '').replace(/\/+$/, '');

  if (hostname.length === 0) {
    hostname = serverHost;
  }

  let result = `${protocol}://${hostname}`;

  if (port) {
    result += `:${port}`;
  }

  if (Array.isArray(path)) {
    path = normalizeAndJoinPath(...path);
  }

  result += path;

  return result;
}

export function prepareURLFromArgs(path, query) {
  const requestUrl = new URL(buildURL());
  requestUrl.pathname = normalizeAndJoinPath(serverBasePath, path);

  for (const param in query) {
    requestUrl.searchParams.set(param, query[param]);
  }

  return requestUrl;
}

async function decodeResponse(response) {
  if (response.ok) {
    const text = await response.text();
    let result = text;

    try {
      result = JSON.parse(text);
    } catch(err) {
      result = text;
    }

    return result;
  }

  return Promise.reject({ status: response.status, message: response.statusText });
}

export async function getFromURLString(urlString) {
  const response = await fetch(urlString, {
    mode: 'cors',
    credentials: 'include',
    headers: getStandardHeaders(false),
  });

  return decodeResponse(response);
}

function buildJSON(body) {
  const sendBody = { ...body };
  delete sendBody.id;
  delete sendBody.url;
  return JSON.stringify(sendBody);
}

export async function postToURLString(urlString, body) {
  const sendJSON = buildJSON(body);
  const response = await fetch(urlString, {
    method: 'POST',
    headers: getStandardHeaders(),
    body: sendJSON,
    mode: 'cors',
    credentials: 'include',
  });

  return decodeResponse(response);
}

export async function putToURLString(urlString, body) {
  const sendJSON = buildJSON(body);
  const response = await fetch(urlString, {
    method: 'PUT',
    headers: getStandardHeaders(),
    body: sendJSON,
    mode: 'cors',
    credentials: 'include',
  });

  return decodeResponse(response);
}