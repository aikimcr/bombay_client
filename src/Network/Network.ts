import { omit } from 'lodash';

// Manage XHR request calls.  This is the low-level stuff.
// import serverConfig from './serverConfig';
export const defaultAPIServer: string = process.env.SERVER;
export const defaultAPIBasePath: string = process.env.SERVER_BASE_PATH;

const getStandardHeaders = (includeContentType = true) => {
  const result: HeadersInit = {};
  if (includeContentType) {
    result['content-type'] = 'application/json';
  }

  const token = localStorage.getItem('jwttoken');

  if (token) {
    result.Authorization = `Bearer ${token}`;
  }

  return result;
};

export const normalizeAndJoinPath = (...pathList: string[]): string => {
  const newPath = pathList.reduce((memo, part) => {
    if (part === undefined) {
      throw new Error(
        `Unable to normalize path with parts: "${pathList.join(', ')}"`,
      );
    }

    let newPart = memo + '/' + part;
    newPart = newPart.replace(/\/\/+/g, '/');

    if (part !== '/') {
      if (newPart.length > 1) {
        newPart = newPart.replace(/\/+$/, '');
      }
    }

    return newPart;
  }, '/');

  return newPath;
};

interface buildUrlProps {
  server?: string;
  basePath?: string;
  applicationPaths?: string[];
}

export const buildURL = ({
  server = defaultAPIServer,
  basePath = defaultAPIBasePath,
  applicationPaths = [],
}: buildUrlProps = {}) => {
  const allPaths = [...applicationPaths];

  if (!server || server === 'default') {
    server = 'http://localhost:2001';
  }

  if (basePath && basePath !== 'default') {
    allPaths.unshift(basePath);
  }

  const result: URL = new URL(server);
  result.pathname = normalizeAndJoinPath(...allPaths);
  return result;
};

export type NetworkURLQueryType = Record<string, string | number | symbol>;

export const prepareURLFromArgs = (
  pathList: string[] = [],
  query: NetworkURLQueryType = {},
) => {
  const requestUrl = buildURL({ applicationPaths: pathList });

  for (const param in query) {
    requestUrl.searchParams.set(param, query[param].toString());
  }

  return requestUrl;
};

const decodeResponse = async (
  response: Awaited<ReturnType<typeof fetch>>,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
): Promise<Record<string, any>> => {
  if (response.ok) {
    const text = await response.text();
    let result = { body: text };

    try {
      result = JSON.parse(text);
    } catch (err) {
      result = { body: text };
    }

    return result;
  }

  return Promise.reject({
    status: response.status,
    message: response.statusText,
  });
};

export const getFromURLString = async (
  urlString: string,
): ReturnType<typeof decodeResponse> => {
  const response = await fetch(urlString, {
    mode: 'cors',
    credentials: 'include',
    headers: getStandardHeaders(false),
  });

  return decodeResponse(response);
};

interface buildJSONProps {
  id?: string | number;
  url?: string;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  [key: string]: any;
}

const buildJSON = (body: buildJSONProps): string => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const sendBody: Record<string, any> = {
    ...omit(body, ['id', 'url']),
  };
  return JSON.stringify(sendBody);
};

export const postToURLString = async (
  urlString: string,
  body: buildJSONProps,
) => {
  const sendJSON = buildJSON(body);
  const response = await fetch(urlString, {
    method: 'POST',
    headers: getStandardHeaders(),
    body: sendJSON,
    mode: 'cors',
    credentials: 'include',
  });

  return decodeResponse(response);
};

export const putToURLString = async (
  urlString: string,
  body: buildJSONProps,
) => {
  const sendJSON = buildJSON(body);
  const response = await fetch(urlString, {
    method: 'PUT',
    headers: getStandardHeaders(),
    body: sendJSON,
    mode: 'cors',
    credentials: 'include',
  });

  return decodeResponse(response);
};
