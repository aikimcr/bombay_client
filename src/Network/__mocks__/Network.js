const RealNetwork = jest.requireActual('../Network');

export const serverProtocol = 'https';
export const serverHost = 'fake';
export const serverBasePath = 'xyzzy';
export const serverPort = null;

export const normalizeAndJoinPath = RealNetwork.normalizeAndJoinPath;

export function buildURL (args = {}) {
    let { hostname, path, protocol, port } = {
        hostname: serverHost,
        path: '/',
        protocol: serverProtocol,
        port: serverPort,
        ...args,
    }

    return RealNetwork.buildURL(args);
}

// ToDo Find a way to do this without having to copy the entire real implementation.
export function prepareURLFromArgs(path, query) {
    const requestUrl = new URL(buildURL());
    requestUrl.pathname = normalizeAndJoinPath(serverBasePath, path);

    for (const param in query) {
        requestUrl.searchParams.set(param, query[param]);
    }

    return requestUrl;
}


export function getFromURLString(url, query) {
    throw new Error(`getFromURLString has not been mocked properly ${url} ${JSON.stringify(query)}`);
}

export async function postToURLString(urlString, body) {
    throw new Error(`postToURLString has not been mocked properly ${url} ${JSON.stringify(body)}`);
}

export async function putToURLString(urlString, body) {
    throw new Error(`putToURLString has not been mocked properly ${url} ${JSON.stringify(body)}`);
}

// Test Utilities not found in the original
let mockPromise;
let mockResolve;
let mockReject;

export function _setupMockPromise() {
    [mockPromise, mockResolve, mockReject] = makeResolvablePromise();
    return {
        promise: mockPromise,
        resolve: mockResolve,
        reject: mockReject,
    };
}

export function _setupMocks() {
    _setupMockPromise();

    this.getFromURLString = jest.fn((urlString, query) => {
        expect(mockPromise).toBeDefined();
        return mockPromise;
    });

    this.postToURLString = jest.fn((urlString, body) => {
        expect(mockPromise).toBeDefined();
        return mockPromise;
    });

    this.putToURLString = jest.fn((urlString, body) => {
        expect(mockPromise).toBeDefined();
        return mockPromise;
    });

    return {
        resolve: mockResolve,
        reject: mockReject,
        promise: mockPromise,
    };
}