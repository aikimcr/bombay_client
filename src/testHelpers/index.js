export * from './ContextChanger';
export * from './modelTools';
