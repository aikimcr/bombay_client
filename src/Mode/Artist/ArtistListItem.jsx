import React, { useState } from 'react';

import './ArtistList.scss';

import { Artist } from './Artist.jsx';
import FormModal from '../../Modal/FormModal.jsx';

function ArtistListItem(props) {
  const [showEdit, setShowEdit] = useState(false);
  const [displayName, setDisplayName] = useState(props.artist.name);

  async function updateArtist(modelDef) {
    await props.artist.set(modelDef).save();
    setDisplayName(props.artist.get('name'));
  }

  return (
    <React.Fragment>
      <li
        className="card"
        data-testid="artist-list-card"
        onClick={() => setShowEdit(true)}
      >
        <div className="header">Artist</div>
        <div className="name">{displayName}</div>
        <div className="details"></div>
      </li>
      <FormModal
        title="Edit Artist"
        onClose={() => setShowEdit(false)}
        onSubmit={updateArtist}
        open={showEdit}
      >
        <Artist artist={props.artist} />
      </FormModal>
    </React.Fragment>
  );
}

export default ArtistListItem;
