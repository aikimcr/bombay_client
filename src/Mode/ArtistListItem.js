import React from 'react';
import { useState } from 'react'

import './ArtistList.scss';

import Artist from './Artist';
import { useFormModal } from '../Modal/FormModal';

function ArtistListItem(props) {
    const [openModal, closeModal, FormModal] = useFormModal();

    async function showArtist() {
        const modelDef = await openModal();
        console.log(modelDef);
    }

    return (
        <React.Fragment>
        <li className="card" onClick={showArtist}>{props.artist.get('name')}</li>
            <FormModal title="Edit Artist">
                <Artist artist={props.artist} />
            </FormModal>
            </React.Fragment>
    )
}

export default ArtistListItem;