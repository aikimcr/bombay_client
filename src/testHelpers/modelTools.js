// Import this after all mocks.  Especially Network and login mocks.

import casual from "casual";

import * as Network from "../Network/Network";

import ModelBase from "../Model/ModelBase";
import ArtistModel from "../Model/ArtistModel";
import SongModel from "../Model/SongModel";

const allNames = {};
casual.define("uniqueName", function (nameType, clear) {
  if (!allNames[nameType]) allNames[nameType] = new Set();

  const oldSize = allNames[nameType].size;
  let newName;

  while (allNames[nameType].size === oldSize) {
    newName = casual.full_name;
    allNames[nameType].add(newName);
  }

  return newName;
});

const idSequences = {};
casual.define("nextId", function (sequenceName) {
  const nextOne = (idSequences[sequenceName] ?? 0) + 1;
  idSequences[sequenceName] = nextOne;
  return nextOne;
});

export function makeAModel(tableName = "table1", fieldsCallback) {
  const def = {};
  def.id = casual.nextId(tableName);
  def.name = casual.uniqueName(tableName);

  if (fieldsCallback) {
    fieldsCallback(def);
  }

  let modelClass = ModelBase;

  switch (tableName) {
    case "artist":
      modelClass = ArtistModel;
      break;

    case "song":
      modelClass = SongModel;
      def.key_signature = "C";
      def.tempo = 120;
      def.lyrics = "O Solo Mio! The troubles I have seen";

      if (def.artist) {
        def.artist_id = def.artist.id;
      } else {
        const [artist] = makeAModel("artist");
        def.artist_id = artist.id;
        def.artist = artist;
      }
      break;

    default:
  }

  def.url = Network.buildURL({ path: `/${tableName}/${def.id}` });

  return [def, modelClass.from(def)];
}

export function makeModels(
  length = 10,
  query = {},
  tableName = "table1",
  fieldsCallback,
) {
  const result = {
    data: [],
  };

  const models = [];

  while (result.data.length < length) {
    const [def, model] = makeAModel(tableName, fieldsCallback);
    result.data.push(def);
    models.push(model);
  }

  let offset = (query.offset || 0) - length;
  let limit = query.limit || length;

  if (offset >= 0) {
    result.prevPage = Network.prepareURLFromArgs(tableName, {
      offset,
      limit,
    }).toString();
  }

  if (limit <= length) {
    offset = (query.offset || 0) + length;
    result.nextPage = Network.prepareURLFromArgs(tableName, {
      offset,
      limit,
    }).toString();
  }

  return [result, models];
}
