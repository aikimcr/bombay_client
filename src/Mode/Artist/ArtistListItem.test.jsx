import React from 'react';
import { act, fireEvent, render, screen } from '@testing-library/react';

import {
  mockGetFromURLString,
  mockPutToURLString,
  mockPostToURLString,
  mockBuildURL,
  mockDefaultAPIBasePath,
  mockDefaultAPIServer,
  mockPrepareURLFromArgs,
} from '../../Network/testing';

jest.mock('../../Network/Network', () => {
  const originalModule = jest.requireActual('../../Network/Network');

  return {
    __esModule: true,
    ...originalModule,
    defaultAPIServer: mockDefaultAPIServer,
    defaultAPIBasePath: mockDefaultAPIBasePath,
    buildURL: mockBuildURL,
    prepareURLFromArgs: mockPrepareURLFromArgs,
    getFromURLString: mockGetFromURLString,
    putToURLString: mockPutToURLString,
    postToURLString: mockPostToURLString,
  };
});

jest.mock('./ArtistEditor/ArtistForm');

import * as Network from '../../Network/Network';

import { makeADef, makeAModelFromDef } from '../../testHelpers/modelTools';
import { TestArtistCollectionURL, TestUrlWithId } from '../../Model/testing';

import { ArtistListItem } from './ArtistListItem';

jest.useFakeTimers();

beforeEach(() => {
  const modalRoot = document.createElement('div');
  modalRoot.id = 'modal-root';
  document.body.append(modalRoot);
});

afterEach(() => {
  const modalRoot = document.getElementById('modal-root');
  modalRoot.remove();
});

it('Component should match snapshot', async () => {
  const modelDef = makeADef('artist', (def) => {
    def.name = 'Alessia Koelpin';
  });
  mockBuildURL.mockReturnValue(
    TestUrlWithId(TestArtistCollectionURL, modelDef.id),
  );
  const model = makeAModelFromDef(modelDef, 'artist');

  const { asFragment } = render(<ArtistListItem artist={model} />);

  expect(asFragment()).toMatchSnapshot();
});

it('should render a list item', async () => {
  const modelDef = makeADef('artist');
  mockBuildURL.mockReturnValue(
    TestUrlWithId(TestArtistCollectionURL, modelDef.id),
  );
  const model = makeAModelFromDef(modelDef, 'artist');
  render(<ArtistListItem artist={model} />);
  expect(screen.getByTestId('artist-list-card')).toBeInTheDocument();
});

it('should open the editor modal', async () => {
  const modelDef = makeADef('artist');
  mockBuildURL.mockReturnValue(
    TestUrlWithId(TestArtistCollectionURL, modelDef.id),
  );
  const model = makeAModelFromDef(modelDef, 'artist');
  render(<ArtistListItem artist={model} />);

  const card = screen.getByTestId('artist-list-card');

  await act(async () => {
    fireEvent.click(card);
  });

  expect(screen.getByTestId('mock-artist-form')).toBeInTheDocument();
});

// Move this to the model editor
it.skip('should save changes to the model', async () => {
  const [mockPromise, mockResolve] = makeResolvablePromise();
  // The mock is not recognized unless it is done this way.
  Network.putToURLString = jest.fn((url, body) => {
    return mockPromise;
  });

  const modelDef = makeADef('artist');

  const result = render(<ArtistListItem artist={model} />);

  const index = result.container;

  await act(async () => {
    index.firstChild.click();
  });

  const modalRoot = document.getElementById('modal-root');
  expect(modalRoot.childElementCount).toEqual(1);

  const submitButton = modalRoot.querySelector('[type="submit"]');

  await changeInput(
    modalRoot.querySelector('[data-targetfield="name"'),
    'input',
    'Herkimer',
    250,
  );

  act(() => {
    submitButton.click();
  });

  await act(async () => {
    mockResolve({ ...modelDef, name: 'Herkimer' });
  });

  expect(Network.putToURLString).toHaveBeenCalledTimes(1);
  expect(Network.putToURLString).toHaveBeenCalledWith(modelDef.url, {
    ...modelDef,
    name: 'Herkimer',
  });
});
