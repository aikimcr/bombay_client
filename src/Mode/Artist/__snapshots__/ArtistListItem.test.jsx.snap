// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Component should match snapshot 1`] = `
<DocumentFragment>
  <li
    class="card"
    data-testid="artist-list-card"
  >
    <div
      class="header"
    >
      Artist
    </div>
    <div
      class="name"
    >
      Alessia Koelpin
    </div>
    <div
      class="details"
    />
    <div
      data-testid="mock-artist-form"
    >
      <p
        data-testid="mock-component-name"
      >
        Name Artist
      </p>
      <div
        data-testid="mock-component-props"
      >
        <div>
          <span>
            name
          </span>
          : 
          <span>
            Artist
          </span>
        </div>
        <div>
          <span>
            classname
          </span>
          : 
          <span>
            mock-artist-form
          </span>
        </div>
        <div>
          <span>
            testId
          </span>
          : 
          <span>
            mock-artist-form
          </span>
        </div>
      </div>
      <div
        data-testid="mock-component-children"
      >
        <p />
        <p>
          Alessia Koelpin
        </p>
      </div>
    </div>
  </li>
</DocumentFragment>
`;
